Setup CI to automatically deploy demo branch to demo.kedro.org

limdauto opened the pull request:
## Description

This PR deploys the demo project to demo.kedro.org with the latest Kedro-Viz version.

## Development notes

<!-- What have you changed? Consider adding a screenshot or GIF. -->

## QA notes

I have tested this manually. Once this is merged, I will test it once more by merging main into demo.

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [ ] Added tests to cover my changes

studioswong reviewed (approved):
I've seen this happen yesterday so all is good!! Perhaps maybe a small update on the release section under our contribution doc about this new deployment setup and infrasture we use would be great. 

Thanks for making this happen! 🤟

limdauto commented:
@studioswong I did add a note :D

studioswong commented:
@limdauto sweet 💯

tynandebold commented on CONTRIBUTING.md:
```diff
@@ -314,3 +314,10 @@ make version VERSION=<version-to-release>
     * Publish the release.
 
 - Once the release tag is published on Github, a CircleCI job will be triggered to push it to npm and PyPI accordingly.
+- Once the new version is on PyPI, you can deploy it to https://demo.kedro.org by merging `main` into `demo` branch. A CI job will automatically build a container using the [demo project](./demo-project) with the newly released version and deploy it.
```
```suggestion
- Once the new version is on PyPI, you can deploy it to https://demo.kedro.org by merging `main` into the `demo` branch. A CI job will automatically build a container using the [demo project](./demo-project) with the newly released version and deploy it.
```

tynandebold commented on .circleci/config.yml:
```diff
@@ -141,31 +144,36 @@ commands:
-            git push origin gh-pages --force
+            echo "kedro_viz==$KEDRO_VIZ_VERSION" >> src/docker_requirements.txt
+            docker build -t $AWS_ECR_URL:$KEDRO_VIZ_VERSION .
+            aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin public.ecr.aws
+            docker push $AWS_ECR_URL:$KEDRO_VIZ_VERSION
+      - run:
+          name: Create a new lightsail deployment
+          command: |
+            # install lightsail cli
+            # run https://docs.aws.amazon.com/cli/latest/reference/lightsail/create-container-service-deployment.html#create-container-service-deployment
+            aws lightsail create-container-service-deployment --region eu-west-2 --cli-input-json file://./lightsail.json
```
Curiosity: why is our ECR registry in `us-east-1` but we deploy it to `eu-west-2`? Am I reading that right? Does it matter? I'm sure it works, just wondering why they're different.

limdauto replied:
AWS only has 1 region for their public registry which is `us-east-1`.

tynandebold commented on .circleci/config.yml:
```diff
@@ -141,31 +144,36 @@ commands:
+          name: Setup environment
+          command: |
+            cd demo-modular-spaceflights
+            # use credentials for kedro-ecr-publish as the credentials in this jobs
```
```suggestion
            # use credentials for kedro-ecr-publish as the credentials in this job
```

tynandebold commented on demo-project/lightsail.json:
```diff
@@ -0,0 +1,15 @@
+  "serviceName": "kedro-viz-live-demo",
+  "containers": {
+    "kedro-viz-live-demo": {
+      "image": "public.ecr.aws/g0x0s3o2/kedro-viz-live-demo:4.2.0",
```
Does `4.2.0` need to change when we release a new version of Viz? That's our current version isn't it?

limdauto replied:
oh yes such a good point. I forgot about this. Let me fix it.

studioswong replied:
indeed great point - that reminds me that our `make version` script needs to update the `.version` file with the latest version as well.

limdauto replied:
@studioswong that actually was done in this PR. I just forgot about this JSON

studioswong replied:
indeed just saw the update the `versioning.py` - I'm blind 🙈 

I tried running the versioning script and it's returning an error due to it referring to the old demo project name - see below comment.

tynandebold reviewed (approved):
Looks great. I'm tentatively approving with some minute suggestions, a couple of questions for my own curiosity, and one more pressing question about the version of Viz used in the `demo-project/lightsail.json` file.

studioswong commented on tools/versioning.py:
```diff
@@ -13,6 +13,7 @@
 VIZ_INIT_FILE = "package/kedro_viz/__init__.py"
 PACKAGE_JSON_FILE = "package.json"
 PACKAGE_JSON_LOCK_FILE = "package-lock.json"
+DEMO_VERSION_FILE = "demo-modular-spaceflights/.version"
```
I think this needs to be renamed back to `demo-project` as the rename was reverted.

studioswong commented on .circleci/config.yml:
```diff
@@ -141,31 +144,36 @@ commands:
+      - run:
+          name: Setup environment
+          command: |
+            cd demo-modular-spaceflights
```
This would need to be updated to `demo-project`

studioswong commented on .circleci/config.yml:
```diff
@@ -141,31 +144,36 @@ commands:
+            echo "AWS_SECRET_ACCESS_KEY=$AWS_ECR_SECRET_ACCESS_KEY" >> $BASH_ENV
+            echo "AWS_ECR_URL=public.ecr.aws/g0x0s3o2/kedro-viz-live-demo" >> $BASH_ENV
+            echo "KEDRO_VIZ_VERSION=$(cat .version)" >> $BASH_ENV
+            echo "cd demo-modular-spaceflights" >> $BASH_ENV
```
same as above

limdauto commented:
@tynandebold @studioswong I have addressed your comments. Either merge this in before the release and rerun the versioning script, or please manually change the version in `demo-pacakge/.version` and `demo-package/lightsail.json` just this one time.

tynandebold commented:
> @tynandebold @studioswong I have addressed your comments. Either merge this in before the release and rerun the versioning script, or please manually change the version in `demo-pacakge/.version` and `demo-package/lightsail.json` just this one time after the release.

I'll merge it now, rerun the versioning script, and then release. That way we'll see if it all works in tandem as we planned on.

limdauto commented:
@tynandebold sure. Any problem please ping me here. Available to fix.

tynandebold commented:
> @tynandebold sure. Any problem please ping me here. Available to fix.

Roger. Stand by...

tynandebold commented:
@limdauto the release is [out](https://github.com/kedro-org/kedro-viz/releases/tag/v4.3.0) now. You should do the honor of merging `main` into `demo` once it hits PyPI.

limdauto commented:
@tynandebold yep, just tested the version on PyPI. Look good. Pushed to demo.

tynandebold commented:
> @tynandebold yep, just tested the version on PyPI. Look good. Pushed to demo.

@limdauto the deploy failed. Let's try and sync up via Discord sometime and fix?

limdauto commented:
@tynandebold looks like AWS credentials isn't working. Might have to do with @idanov rotating credentials for `kedro-ecr-publish`

limdauto commented:
@tynandebold I have just checked. CircleCI for Kedro-Viz is still using the old key. You can ask Ivan to change the key to new one (identical to the ones used on Kedro atm) and we should be good. I don't have access to those keys anymore.

tynandebold commented:
@limdauto we did that with no luck.

We're still getting a security token error when running the [Dockerfile](https://github.com/kedro-org/kedro-viz/blob/main/demo-project/Dockerfile#L10).

```
An error occurred (UnrecognizedClientException) when calling the GetAuthorizationToken operation: The security token included in the request is invalid.
```

Can you write to me on Discord sometime today? Username: tySnipes5150#6537.
